### ISSUE TITLE
Consider enforcing the Developer's Certificate of Origin (DCO, `Signed-off-by:`)

### ISSUE THREAD

**Issue by qmonnet** [githedgehog/dataplane#146]
Following internal discussions, I would like to promote the use, and the enforcement, of the Developer's Certificate of Origin (DCO) for the code in this repository.

Here's [the text of the certificate](https://developercertificate.org/):

> Developer's Certificate of Origin 1.1
>
> By making a contribution to this project, I certify that:
>
> (a) The contribution was created in whole or in part by me and I
>     have the right to submit it under the open source license
>     indicated in the file; or
>
> (b) The contribution is based upon previous work that, to the best
>     of my knowledge, is covered under an appropriate open source
>     license and I have the right under that license to submit that
>     work with modifications, whether created in whole or in part
>     by me, under the same open source license (unless I am
>     permitted to submit under a different license), as indicated
>     in the file; or
>
> (c) The contribution was provided directly to me by some other
>     person who certified (a), (b) or (c) and I have not modified
>     it.
>
> (d) I understand and agree that this project and the contribution
>     are public and that a record of the contribution (including all
>     personal information I submit with it, including my sign-off) is
>     maintained indefinitely and may be redistributed consistent with
>     this project or the open source license(s) involved.

I believe all four points are prerequisites for accepting contributions, either from Hedgehog team members or from external contributors.

Concretely, this comes down to signing off every commit. As an author, or a contributor to a given commit, I would add the tag `Signed-off-by: Full Name <email@address>`, with my name and email, to the end of the commit description. This is made simple with the relevant Git option: `git commit -s` will add the sign-off using my user name and email. There are ways to automate it from Git's configuration file, although I personally don't recommend automating the sign-offs, because signing off a commit should remain a conscious commitment.

For an example of signed-off commits, look at [my commits](https://github.com/githedgehog/dataplane/commit/bc2631f44b27464399528c761b41e28e38946862) in this repository.

Enforcing the use of DCO for contributions is easy to do in CI, for example with the relevant Action from GitHub: https://github.com/apps/dco.

“Thousands of projects” (according to CNCF) rely on and enforce the use of DCO. Notable examples include the Linux kernel or Cilium. The CNCF has some [additional guidelines on the topic](https://github.com/cncf/foundation/blob/main/dco-guidelines.md), also providing more details for the motivation. Alternatives to the DCO include signing Contributor License Agreements (CLAs) before contributing to a project, but CLAs are a heavier process.

Risks:

* Moderate slow-down of the time-to-merge for Pull Requests, possibly some frustration, until contributors get used to signing off their commits.

Benefits:

* Better legal coverage for accepting external contributions (_I am not a lawyer_).

The idea would be to extend the use of DCO to other user-facing Hedgehog repositories as well; I'm starting with the one I'm familiar with.

Process:

1. Discuss with the team and make sure we're on board with using and enforcing DCO in general
2. Add the GitHub application to the repo to actually enforce it
3. Add the GitHub application to other repositories

**Comment by daniel-noland** [comment]
I think it is fair to say that this is implemented in dataplane now.  Close as done?

**Comment by qmonnet** [comment]
Yes. Sergei enabled the GH Action organization-wide, enforcement is a matter of marking it as required for each repo. This is the case for dataplane, we can close.
